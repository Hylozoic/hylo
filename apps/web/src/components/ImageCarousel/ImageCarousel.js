import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import Slider from 'react-slick'
import { filter, isEmpty } from 'lodash/fp'

import 'slick-carousel/slick/slick.css'
import 'slick-carousel/slick/slick-theme.css'
import classes from './ImageCarousel.module.scss'

export default function ImageCarousel ({
  attachments,
  initialSlide = 0
}) {
  const imageAttachments = filter({ type: 'image' }, attachments)

  const settings = {
    dots: true,
    infinite: true,
    speed: 500,
    slidesToShow: 1,
    arrows: imageAttachments.length > 1,
    slidesToScroll: 1,
    adaptiveHeight: true,
    initialSlide: parseInt(initialSlide)
  }

  const carouselRef = useRef()
  const slickRef = useRef()

  const handleKeydown = event => {
    if (event.code === 'ArrowLeft') slickRef.current.slickPrev()
    if (event.code === 'ArrowRight') slickRef.current.slickNext()
  }

  if (isEmpty(imageAttachments)) return null

  return (
    <div className={classes.images} ref={carouselRef} onKeyDown={handleKeydown}>
      <Slider ref={slickRef} {...settings}>
        {imageAttachments.map((image, index) =>
          <div className={classes.imageWrapper} key={index} data-testid={`sc-img${index}`}>
            <img
              src={image.url}
              alt={`Attached image ${index + 1}`}
            />
          </div>
        )}
      </Slider>
    </div>
  )
}

ImageCarousel.propTypes = {
  attachments: PropTypes.array.isRequired
}
