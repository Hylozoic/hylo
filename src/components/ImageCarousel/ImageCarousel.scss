.images {
  :global(.slick-prev:before),
  :global(.slick-next:before) {
    text-shadow: 0px 0px 5px black;
  }

  :global(.slick-list) {
    margin-bottom: 15px;
  }

  :global(.slick-dots) {
    bottom: -20px;
  }
}