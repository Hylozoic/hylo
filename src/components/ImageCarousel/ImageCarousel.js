import React from 'react'
import PropTypes from 'prop-types'
import Slider from 'react-slick'
import { filter, isEmpty } from 'lodash/fp'

import 'slick-carousel/slick/slick.css'
import 'slick-carousel/slick/slick-theme.css'
import './ImageCarousel.scss'

export default function ImageCarousel ({
  attachments,
  initialSlide = 0
}) {
  const imageAttachments = filter({ type: 'image' }, attachments)
  if (isEmpty(imageAttachments)) return null

  const settings = {
    dots: true,
    infinite: true,
    speed: 500,
    slidesToShow: 1,
    arrows: true,
    slidesToScroll: 1,
    adaptiveHeight: true,
    initialSlide: parseInt(initialSlide)
  }

  return (
    <div styleName='images'>
      <Slider {...settings}>
        {imageAttachments.map((image, index) =>
          <img
            src={image.url}
            alt={`Attached image ${index + 1}`}
            key={index}
            data-testid={`sc-img${index}`}
          />
        )}
      </Slider>
    </div>
  )
}

ImageCarousel.propTypes = {
  attachments: PropTypes.array.isRequired
}
