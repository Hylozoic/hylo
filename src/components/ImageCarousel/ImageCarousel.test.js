import React from 'react'
import ImageCarousel from './ImageCarousel'
import { shallow } from 'enzyme'

it('renders a single image', () => {
  expect(shallow(<ImageCarousel attachments={[
    { url: 'foo', type: 'image' },
    { url: 'bonkerz', type: 'file' }
  ]} />)).toMatchSnapshot()
})

it('renders multiple images', () => {
  expect(shallow(<ImageCarousel attachments={[
    { url: 'bar', type: 'image' },
    { url: 'baz', type: 'image' },
    { url: 'bonk', type: 'image' },
    { url: 'bonkerz', type: 'file' }
  ]} />)).toMatchSnapshot()
})
