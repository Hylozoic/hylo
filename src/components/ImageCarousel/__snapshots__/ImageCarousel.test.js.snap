// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a single image 1`] = `
<div
  data-stylename="images"
>
  <Slider
    adaptiveHeight={true}
    arrows={true}
    dots={true}
    infinite={true}
    initialSlide={0}
    slidesToScroll={1}
    slidesToShow={1}
    speed={500}
  >
    <img
      alt="Attached image 1"
      data-testid="sc-img0"
      key="0"
      src="foo"
    />
  </Slider>
</div>
`;

exports[`renders multiple images 1`] = `
<div
  data-stylename="images"
>
  <Slider
    adaptiveHeight={true}
    arrows={true}
    dots={true}
    infinite={true}
    initialSlide={0}
    slidesToScroll={1}
    slidesToShow={1}
    speed={500}
  >
    <img
      alt="Attached image 1"
      data-testid="sc-img0"
      key="0"
      src="bar"
    />
    <img
      alt="Attached image 2"
      data-testid="sc-img1"
      key="1"
      src="baz"
    />
    <img
      alt="Attached image 3"
      data-testid="sc-img2"
      key="2"
      src="bonk"
    />
  </Slider>
</div>
`;
